### ISSUE TITLE
"Serialization of 'Closure' is not allowed" when trying to cache query from Similar

### ISSUE THREAD

**Issue by aelvan** [craftcms/cms#2674]
### Description
Seems like there are still some issues with serialization of element queries, as reported in [issue 2586](https://github.com/craftcms/cms/issues/2586). When trying to cache template code that uses the [Similar plugin](https://github.com/nystudio107/craft-similar) I get the same error on the latest RC. Although the error is triggered by a plugin, I think the root cause has to be fixed in core. 

The offending code is this ([line 90 in /services/Similar.php](https://github.com/nystudio107/craft-similar/blob/v1/src/services/Similar.php#L90)):

```
      $query->on(ElementQuery::EVENT_AFTER_PREPARE, function(CancelableEvent $event) {
            /** @var ElementQuery $query */
            $query = $event->sender;
            // Add in the `count` param so we know how many were fetched
            $query->query->addSelect(['COUNT(*) as count']);
            $query->query->orderBy('count DESC, ' . str_replace('`', '', $this->preOrder));
            $query->query->groupBy('{{%relations}}.sourceId');
            $event->isValid = true;
        });
```

The problem is the anonymous function in the event handler. Refactoring to this fixes the problem:

```
    $query->on(ElementQuery::EVENT_AFTER_PREPARE, [$this, 'eventAfterPrepareHandler']);

    ...

    protected function eventAfterPrepareHandler(CancelableEvent $event) {
        /** @var ElementQuery $query */
        $query = $event->sender;
        // Add in the `count` param so we know how many were fetched
        $query->query->addSelect(['COUNT(*) as count']);
        $query->query->orderBy('count DESC, ' . str_replace('`', '', $this->preOrder));
        $query->query->groupBy('{{%relations}}.sourceId');
        $event->isValid = true;
    }
```

But, using anonymous functions in the event handler seems like something that should be possible.

### Steps to reproduce

1. Enable template caching
2. Install Similar, nystudio107/craft-similar
3. Create a section and a category group. Create a category field and add it to the entry type.
4. Create a couple of entries with relations to different categories.
5. In an entry template, do:

```
{% cache %}
    {% set similarEntries = craft.similar.find({ element: entry, context: entry.categories }) %}
{% endcache %}

```

Log with stack trace is attached.

### Additional info

- Craft version: RC 17.1
- PHP version: 7.0.27
- Database driver & version: Mysql 5.6.38
- Plugins & versions: Similar 1.0.0

[serialization_error_log.log](https://github.com/craftcms/cms/files/1875063/serialization_error_log.log)

**Comment by brandonkelly** [comment]
Agree it’s not ideal. Not sure what a better option is though.

**Comment by khalwat** [comment]
I ended up fixing this in Similar 1.0.1 by refactoring it out to a function as you mentioned @aelvan -- but agree it probably should be addressed in core at some point.

**Comment by brandonkelly** [comment]
Most likely it will be addressed by us just removing the element query caching feature/headache altogether.

**Comment by brandonkelly** [comment]
This is now resolved for Craft 3.5. See https://github.com/craftcms/cms/issues/1507#issuecomment-633147835 for details.
